### PR TITLE
[shared_preferences] Adds macOS support with example app

### PR DESCRIPTION
**Pull request by franciscojma86**
## Description

Adds the macos implementation of the shared_preferences plugin.
Tested by adding a example app that runs the e2e tests.

## Related Issues

Fixes flutter/flutter#41718

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/shared_preferences/shared_preferences_web/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.1.0+1
+
+* Adds `package:shared_preferences_macos`.
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences_web/CHANGELOG.md]:
```suggestion
* Increase minimum Flutter SDK to 1.12.8.
```

Would it make sense to do this change in a separate PR?

File: packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## 1.0.1
+
+* Adds `package:shared_preferences_macos`.
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md]:
```suggestion
* Increase minimum Flutter SDK to 1.12.8.
```

Would it make sense to do this change in a separate PR?

File: packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import 'dart:async';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:e2e/e2e.dart';
+
+void main() {
+  E2EWidgetsFlutterBinding.ensureInitialized();
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart]: I think that e2e tests should live in the app-facing plugin, not the platform implementation. Otherwise they'll be duplicated for every platform.

File: packages/shared_preferences/shared_preferences_macos/example/lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+void main() {
+  runApp(MyApp());
+}
+
+class MyApp extends StatelessWidget {
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences_macos/example/lib/main.dart]: Do we need to have this app? It seems like it would be preferable to re-use the app-facing plugin's example app?

File: packages/shared_preferences/shared_preferences_macos/example/android/app/src/main/java/io/flutter/plugins/sharedpreferencesexample/MainActivityTest.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugins.sharedpreferencesexample;
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences_macos/example/android/app/src/main/java/io/flutter/plugins/sharedpreferencesexample/MainActivityTest.java]: With the latest cherry-pick (v1.12.13+hotfix.4), I'm not sure we need Android files.

https://github.com/flutter/flutter/pull/46632

File: packages/shared_preferences/shared_preferences/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+## 0.5.4+9
+
+* Adds `package:shared_preferences_macos`.
```
Comment (collinjackson) [inline packages/shared_preferences/shared_preferences/CHANGELOG.md]:
```suggestion
* Update pubspec.yaml with federated plugin syntax.
* Update minimum Flutter SDK to 1.12.8.
```

File: packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart (hunk 3 above)
Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart]: But these are the only test we have for the package. IIUC, If somebody edits the package and breaks it, the app-facing plugin wouldn't detect it, is that right? 

If we don't need them here, then we don't need the example app at all since that was it's only purpose.

Comment (collinjackson) [reply packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart]: I think it makes sense to have the app-facing plugin's example app include every endorsed platform, not just iOS and Android.

For plugins that have lightweight e2e integration tests like this one, we can teach our CI to run the them using the app-facing plugin's example app for every endorsed platform implementation, using a path dependency. That doesn't need to happen to land this PR, but I think we should file an issue and start heading in that direction.

We can also have heavyweight integration tests involving a native UI / lifecycle interaction, and in fact I'm planning to land an Espresso test of shared_preferences as soon as https://github.com/flutter/plugins/pull/2369 lands. I think this test should still live in the app-facing plugin though, so it can share the Dart code of the example app.

@amirh may have opinions about this as well

Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart]: Yeah, this pr adds macos testing on the app-facing plugin. https://github.com/flutter/plugins/pull/2386 This example app was suggested by @amirh to be able to test the macos package and be "self contained". I'm wondering if we need them since once that pr is landed, any changes to the macos plugin will trigger the app-facing tests which will run the macos implementation.

File: packages/shared_preferences/shared_preferences_macos/CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+## 0.0.1
+
+* Initial Open Source release.
```
[the remark this answers is not in the document]
Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_macos/CHANGELOG.md]: I think I copied it from the web implementation, but i agree it's weird capitalized

File: packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md (hunk 2 above)
Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md]: This was required by the CI tests. Failing if the version wasn't increased.

File: packages/shared_preferences/shared_preferences_macos/example/android/app/src/main/java/io/flutter/plugins/sharedpreferencesexample/MainActivityTest.java (hunk 5 above)
Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_macos/example/android/app/src/main/java/io/flutter/plugins/sharedpreferencesexample/MainActivityTest.java]: I think that's for dummy android apps inside shared_preferences_macos/android, but this is the android part of the example app. But yeah, we have that here too so I can remove it now.

Done

File: packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md (hunk 2 above)
Comment (franciscojma86) [reply packages/shared_preferences/shared_preferences_platform_interface/CHANGELOG.md]: Sorry, another PR was landed before this one, changing the version and changelog covering your suggestion. I still had to bump the version and the only change was adding the macos package.

File: packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart (hunk 3 above)
Comment (collinjackson) [reply packages/shared_preferences/shared_preferences_macos/example/test_driver/shared_preferences_e2e.dart]: This is fine for now but let's figure out how to share example app code in the future. It seems like we may be able to do it with dependency overrides.

Filed https://github.com/flutter/flutter/issues/46716 (maybe we can reference that in a TODO?)
